src/engraving/editing/cmd.cpp

```diff
@@ -3304,45 +3306,57 @@ void Score::cmdMirrorNoteHead()
 
 void Score::cmdIncDecDuration(int nSteps, bool stepDotted)
 {
-    EngravingItem* el = selection().element();
-    if (el == 0) {
-        return;
-    }
-    if (el->isNote()) {
-        el = el->parentItem();
-    }
-    if (!el->isChordRest()) {
-        return;
-    }
-
-    ChordRest* cr = toChordRest(el);
-
-    // if measure rest is selected as input, then the correct initialDuration will be the
-    // duration of the measure's time signature, else is just the ChordRest's duration
-    TDuration initialDuration = cr->durationType();
-    if (initialDuration == DurationType::V_MEASURE) {
-        initialDuration = TDuration(cr->measure()->timesig(), true);
-
-        if (initialDuration.fraction() < cr->measure()->timesig() && nSteps > 0) {
-            // Duration already shortened by truncation; shorten one step less
-            --nSteps;
+    if (m_selection.isRange()) {
+        if (!m_selection.canCopy()) {
+            return;
+        }
+        ChordRest* firstCR = m_selection.firstChordRest();
+        if (firstCR->isGrace()) {
+            firstCR = toChordRest(firstCR->parent());
+        }
+        TDuration initialDuration = firstCR->ticks();
+        TDuration d = initialDuration.shiftRetainDots(nSteps, stepDotted);
+        if (!d.isValid()) {
+            return;
+        }
+        Fraction scale = d.ticks() / initialDuration.ticks();
+        for (ChordRest* cr : getSelectedChordRests()) {
+            Fraction newTicks = cr->ticks() * scale;
+            if (newTicks < Fraction(1, 1024)
+                || (stepDotted && cr->durationType().dots() != firstCR->durationType().dots()
+                    && !cr->isGrace())) {
+                return;
+            }
+        }
+        const muse::ByteArray mimeData(m_selection.mimeData());
+        XmlReader e(mimeData);
+        deleteRange(m_selection.startSegment(), m_selection.endSegment(), staff2track(m_selection.staffStart()),
+                    staff2track(m_selection.staffEnd()), selectionFilter(), m_selection.rangeContainsMultiNoteChords());
+        pasteStaff(e, m_selection.startSegment(), m_selection.staffStart(), scale);
+    } else if (m_selection.isList()) {
+        const std::set<ChordRest*> crs = getSelectedChordRests();
+        for (ChordRest* cr : crs) {
+            TDuration newDuration(stepDotted
+                                  ? cr->durationType().fraction() * Fraction(3, 3 + nSteps)
+                                  : cr->durationType().fraction()* Fraction(3 - nSteps, 3 + nSteps), true);
```

Comment:
Should bring back measure-rest check.
And, replacing the fractions part with shift functions gives:

```suggestion
            // if measure rest is selected as input, then the correct initialDuration will be the
            // duration of the measure's time signature, else is just the ChordRest's duration
            TDuration initialDuration = cr->durationType();
            if (initialDuration == DurationType::V_MEASURE) {
                initialDuration = TDuration(cr->measure()->timesig(), true);

                if (initialDuration.fraction() < cr->measure()->timesig() && nSteps > 0) {
                    // Duration already shortened by truncation; shorten one step less
                    --nSteps;
                }
            }

            TDuration newDuration(stepDotted ? initialDuration.shiftRetainDots(nSteps, stepDotted) : initialDuration.shift(nSteps));
```